app/src/main/java/org/mozilla/firefox/vpn/apptunneling/ui/ExpandableAdapter.kt

```diff
@@ -0,0 +1,167 @@
+        }
+        notifyItemChanged(holder.adapterPosition)
+    }
+
+    private fun onAppGroupActionClicked(holder: AppGroupViewHolder) {
+        val appGroupItem = items[holder.adapterPosition] as AppGroup
+
+        if (appGroupItem.appItems.isEmpty()) return
+
+        val packageNameSet = appGroupItem.appItems.map { it.applicationInfo.packageName }.toSet()
+
+        if (appGroupItem.type == AppGroupType.UNPROTECTED) {
+            expandableItemCallback.onProtectAllClicked(packageNameSet)
+        } else {
+            expandableItemCallback.onUnprotectAllClicked(packageNameSet)
+        }
+    }
+
+    private fun onAppItemActionClicked(holder: AppItemViewHolder) {
+        val appItem = items[holder.adapterPosition] as AppItem
+
+        if (appItem.type == AppGroupType.UNPROTECTED) {
+            expandableItemCallback.onUnprotectedAppChecked(appItem.applicationInfo)
+        } else {
+            expandableItemCallback.onProtectedAppChecked(appItem.applicationInfo)
+        }
+    }
+
+    private fun AppTunnelingUiModel.toExpandableList(): MutableList<ExpandableItem> {
+        val list = mutableListOf<ExpandableItem>()
+        val (exclude, include) = packageList.partition { excludeList.contains(it.packageName) }
+        val excludeAppItems = exclude.map { AppItem(AppGroupType.UNPROTECTED, it) }
+        val includeAppItems = include.map { AppItem(AppGroupType.PROTECTED, it) }
+
+        list.add(AppGroup(AppGroupType.UNPROTECTED, excludeAppItems))
+        list.addAll(excludeAppItems)
+        list.add(AppGroup(AppGroupType.PROTECTED, includeAppItems))
+        list.addAll(includeAppItems)
+        return list
+    }
+
+    private fun getDiffResult(oldList: List<ExpandableItem>, newList: List<ExpandableItem>): DiffUtil.DiffResult {
+
+        return DiffUtil.calculateDiff(object : DiffUtil.Callback() {
+
+            override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
+                val newItem = newList[newItemPosition]
+                val oldItem = oldList[oldItemPosition]
+                if (newItem is AppGroup && oldItem is AppGroup) {
+                    return newItem.type == oldItem.type
+                }
+                if (newItem is AppItem && oldItem is AppItem) {
+                    return newItem.applicationInfo.packageName == oldItem.applicationInfo.packageName
+                }
+                return false
+            }
+
+            override fun getOldListSize(): Int {
+                return oldList.size
+            }
+
+            override fun getNewListSize(): Int {
+                return newList.size
+            }
+
+            override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
+                val newItem = newList[newItemPosition]
+                val oldItem = oldList[oldItemPosition]
+                if (newItem is AppGroup && oldItem is AppGroup) {
+                    return newItem.appItems == oldItem.appItems &&
```

Comment:
Not sure, but I believe `newItem.appItems.size == oldItem.appItems.size` is suffient here, no need to check the content and order of the lists